tensorboard/plugins/debugger/numerics_alert.py

```diff
@@ -0,0 +1,338 @@
+        history.add(numerics_alert)
+        self._data[key] = history
+
+  def report(self, device_name_filter=None, tensor_name_filter=None):
+    """Get a report of offending device/tensor names.
+
+    The report includes information about the device name, tensor name, first
+    (earliest) timestamp of the alerting events from the tensor, in addition to
+    counts of nan, positive inf and negative inf events.
+
+    Args:
+      device_name_filter: regex filter for device name, or None (not filtered).
+      tensor_name_filter: regex filter for tensor name, or None (not filtered).
+
+    Returns:
+      A list of NumericsAlertReportRow, sorted by the first_timestamp in
+      asecnding order.
+    """
+    report = []
+    for key in self._data:
+      device_name, tensor_name = key
+      history = self._data[key]
+      report.append(
+          NumericsAlertReportRow(
+              device_name=device_name,
+              tensor_name=tensor_name,
+              first_timestamp=history.first_timestamp(),
+              nan_event_count=history.event_count(constants.NAN_KEY),
+              neg_inf_event_count=history.event_count(constants.NEG_INF_KEY),
+              pos_inf_event_count=history.event_count(constants.POS_INF_KEY)))
+
+    if device_name_filter:
+      device_name_pattern = re.compile(device_name_filter)
+      report = [item for item in report
+                if device_name_pattern.match(item.device_name)]
+    if tensor_name_filter:
+      tensor_name_pattern = re.compile(tensor_name_filter)
+      report = [item for item in report
+                if tensor_name_pattern.match(item.tensor_name)]
+    # Sort results chronologically.
+    return sorted(report, key=lambda x: x.first_timestamp)
+
+  def create_jsonable_registry(self):
+    """Creates a JSON-able representation of this object.
+
+    Returns:
+      A dictionary mapping (device, tensor name) to JSON-able object
+      representations of NumericsAlertHistory.
+    """
+    # JSON does not support tuples as keys. Only strings. Therefore, we store
+    # the device name, tensor name, and dictionary data within a 3-item list.
+    return [HistoryTriplet(pair[0], pair[1], history.create_jsonable_history())
+            for (pair, history) in self._data.items()]
+
+
+def extract_numerics_alert(event):
+  """Determines whether a health pill event contains bad values.
+
+  A bad value is one of NaN, -Inf, or +Inf.
+
+  Args:
+    event: (`Event`) A `tensorflow.Event` proto from `DebugNumericSummary`
+      ops.
+
+  Returns:
+    An instance of `NumericsAlert`, if bad values are found.
+    `None`, if no bad values are found.
+
+  Raises:
+    ValueError: if the event does not have the expected tag prefix or the
+      debug op name is not the expected debug op name suffix.
+  """
+  value = event.summary.value[0]
+  debugger_plugin_metadata_content = None
+  if value.HasField("metadata"):
+    for plugin_data in value.metadata.plugin_data:
```

Comment:
Again, make sure that https://github.com/tensorflow/tensorflow/pull/11952 is on your radar.